drs_downloader/manager.py

```diff
@@ -0,0 +1,383 @@
+                               disable=self.disable)
+        ]
+
+        # second, download the parts
+        tasks = []
+        for drs_object in drs_objects_with_signed_urls:
+            task = asyncio.create_task(
+                self._run_download_parts(drs_object=drs_object, destination_path=destination_path))
+            tasks.append(task)
+
+        drs_objects_with_file_parts = [
+            await f
+            for f in asyncio.as_completed(tasks)
+        ]
+
+        return drs_objects_with_file_parts
+
+    async def _run_sign_urls(self, drs_objects: List[DrsObject], leave: bool) -> List[DrsObject]:
+        """Create asyncio tasks to retrieve signed DrsObject, displays progress.
+
+        Args:
+            drs_objects: list of drs objects to sign
+            leave: controls progress bar
+
+        Returns:
+            updated list of drs objects
+
+        """
+        tasks = []
+        for drs_object in drs_objects:
+            task = asyncio.create_task(self._drs_client.sign_url(drs_object=drs_object))
+            tasks.append(task)
+
+        updated_drs_objects = [
+            await f
+            for f in tqdm.tqdm(asyncio.as_completed(tasks), total=len(tasks), leave=leave, desc="signing batch",
+                               disable=self.disable)
+        ]
+
+        return updated_drs_objects
+
+    async def _run_get_objects(self, object_ids: List[str], leave: bool) -> List[DrsObject]:
+        """Create async tasks to retrieve list DrsObject, displays progress.
+
+        Args:
+            object_ids: object_id from manifest
+            leave: leave flag to keep progress indicator displayed
+
+        Returns:
+
+        """
+
+        tasks = []
+        for object_id in object_ids:
+            task = asyncio.create_task(self._drs_client.get_object(object_id=object_id))
+            tasks.append(task)
+
+        signed_urls = [
+            await f
+            for f in tqdm.tqdm(asyncio.as_completed(tasks), total=len(tasks), leave=leave, desc="  * batch",
+                               disable=self.disable)
+        ]
+
+        return signed_urls
+
+    @classmethod
+    def _chunker(cls, seq: Iterator, size: int) -> Iterator:
+        """Iterate over a list in chunks.
+
+        Args:
+            seq: an iterable
+            size: desired chunk size
+
+        Returns:
+            an iterator that returns lists of size or less
+        """
+        return (seq[pos:pos + size] for pos in range(0, len(seq), size))
+
+    def get_objects(self, object_ids: List[str]) -> List[DrsObject]:
+        """Create tasks for all object_ids, run them in batches, get information about the object.
+
+        Args:
+            object_ids: list of objects to fetch
+        """
+
+        drs_objects = []
+
+        total_batches = len(object_ids) / self.max_simultaneous_object_retrievers
+        # if fractional
+        if total_batches - round(total_batches) > 0:
```

Comment:
This was talked about yesterday but Math.ciel works better here since if Object_ids/MAX_RETRIEVERS equals a number with any kind of remainder you would need another batch to account for the extra files. round() would cut anything less than half a batch off, leaving files not being downloaded.